### PR TITLE
Added a new generate:command command

### LINKED ISSUE #332: [Proposal] Generate a command class quickly !
Hi all ! 

What do you think about a command to quickly generate a command class ? 
If we have to create a command line app , it's annoying to manually create each time a class for a command.

I propose something like `generate:command --command=AppBundle:Acme`. This will result in the creation of a class named `AcmeCommand` in `Command/` folder inside `AppBundle`.

Please let me know what do you think !

### PR DESCRIPTION
**Pull request by javiereguiluz**
This fixes #332.

If you execute the following:

```bash
$ php app/console generate:command AppBundle app:publish-posts
```

This command generates the following class:

```php
<?php

namespace AppBundle\Command;

use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;

class AppPublishPostsCommand extends ContainerAwareCommand
{
    protected function configure()
    {
        $this
            ->setName('app:publish-posts')
            ->setDescription('...')
            ->addArgument('argument', InputArgument::OPTIONAL, 'Argument description')
            ->addOption('option', null, InputOption::VALUE_NONE, 'Option description')
        ;
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $argument = $input->getArgument('argument');

        if ($input->getOption('option')) {
            // ...
        }

        $output->writeln('Command result.');
    }

}
```

**Note**  there ae two tests failing. I need help to solve this problem. Thanks!

### CODE REVIEW

**Comment by xabbuh** [comment]
@javiereguiluz Which tests are you talking about?

**Comment by javiereguiluz** [comment]
@xabbuh these are the failing tests:

```
There were 2 failures:

1) Sensio\Bundle\GeneratorBundle\Tests\Command\GenerateCommandCommandTest::testInteractiveCommand with data set #0 (array(), 'FooBarBundle\napp:my-command\n', array('FooBarBundle', 'app:foo-bar'))
Expectation failed for method name is equal to <string:generate> when invoked 1 time(s)
Parameter 1 for invocation Sensio\Bundle\GeneratorBundle\Generator\CommandGenerator::generate(Mock_BundleInterface_80d169f2 Object (...), null) does not match expected value.
Failed asserting that null matches expected 'app:foo-bar'.

/Users/javier/sf/bundles/SensioGeneratorBundle/Command/GenerateCommandCommand.php:156
/Users/javier/sf/bundles/SensioGeneratorBundle/vendor/symfony/console/Command/Command.php:259
/Users/javier/sf/bundles/SensioGeneratorBundle/vendor/symfony/console/Tester/CommandTester.php:80
/Users/javier/sf/bundles/SensioGeneratorBundle/Tests/Command/GenerateCommandCommandTest.php:39

2) Sensio\Bundle\GeneratorBundle\Tests\Command\GenerateCommandCommandTest::testInteractiveCommand with data set #1 (array('FooBarBundle'), 'app:foo-bar\n', array('FooBarBundle', 'app:foo-bar'))
Expectation failed for method name is equal to <string:generate> when invoked 1 time(s)
Parameter 1 for invocation Sensio\Bundle\GeneratorBundle\Generator\CommandGenerator::generate(Mock_BundleInterface_80d169f2 Object (...), null) does not match expected value.
Failed asserting that null matches expected 'app:foo-bar'.

/Users/javier/sf/bundles/SensioGeneratorBundle/Command/GenerateCommandCommand.php:156
/Users/javier/sf/bundles/SensioGeneratorBundle/vendor/symfony/console/Command/Command.php:259
/Users/javier/sf/bundles/SensioGeneratorBundle/vendor/symfony/console/Tester/CommandTester.php:80
/Users/javier/sf/bundles/SensioGeneratorBundle/Tests/Command/GenerateCommandCommandTest.php:39
```

**Comment by javiereguiluz** [comment]
All errors have been fixed. This PR is ready for review. Thanks!

File: Command/GenerateCommandCommand.php [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        $this
+            ->setName('generate:command')
+            ->setDescription('Generates a console command')
+            ->setDefinition(array(
+                new InputArgument('bundle', InputArgument::OPTIONAL, 'The bundle where the controller is generated'),
+                new InputArgument('name', InputArgument::OPTIONAL, 'The name of the command as you type it in the console (e.g. app:my-command)'),
+            ))
+            ->setHelp(<<<EOT
+The <info>generate:command</info> command helps you generates new commands
```
Comment (xabbuh) [inline Command/GenerateCommandCommand.php]: [...] helps you to generate [...]

File: Command/GenerateCommandCommand.php [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+inside bundles. Provide the bundle name as the first argument and the command
+name as the second argument:
+
+<info>php app/console generate:command AcmeBlogBundle blog:publish-posts</info>
+
+If any of the arguments is missing, the command will ask for their values
+interactively. If you want to disable any user interaction, use
+<comment>--no-interaction</comment> but don't forget to pass all needed arguments.
```
Comment (xabbuh) [inline Command/GenerateCommandCommand.php]: missing comma before "but"

File: Command/GenerateCommandCommand.php [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+Every generated file is based on a template. There are default templates but they can
```
Comment (xabbuh) [inline Command/GenerateCommandCommand.php]: same here

File: Command/GenerateCommandCommand.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+be overriden by placing custom templates in one of the following locations, by order of priority:
+
+<info>BUNDLE_PATH/Resources/SensioGeneratorBundle/skeleton/command
+APP_PATH/Resources/SensioGeneratorBundle/skeleton/command</info>
+
+You can check https://github.com/sensio/SensioGeneratorBundle/tree/master/Resources/skeleton
+in order to know the file structure of the skeleton
```
Comment (xabbuh) [inline Command/GenerateCommandCommand.php]: missing period

**Comment by javiereguiluz** [comment]
Errors reported by @xabbuh have been fixed. Any other comment about this new code before considering it finished? Thanks.

**Comment by OskarStark** [comment]
Thanks for your work.

Are you sure you want ```ContainerAwareCommand``` every time you create a command?

Oskar

**Comment by javiereguiluz** [comment]
@OskarStark good question! First, we don't want to define an option for this :)  So we must make a decision. This is the reasoning we followed:

If we extend from `Command`, some people will have problems if they try to use `$this->container->...` as showed in Symfony docs and blogs. However, if we extend from `ContainerAwareCommand`, all developers will be happy: if you try to use the container, everything will work. If you don't use the container, this doesn't make your command fail.

**Comment by OskarStark** [comment]
thats exactly what i thought, and i absolutely agree with you!

i just wanted to hint you to that! :)

**Comment by xabbuh** [comment]
I agree that extending the `ContainerAwareCommand` class is the better solution. Most of the times you will probably need to access some services in your command and when you register the command itself as a service, you will probably not use the generator at all for this task.

**Pull request closed without merging**